clients/unshaded/src/main/java/tachyon/client/block/BSContext.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.client.block;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import com.google.common.base.Preconditions;
+
+import tachyon.client.BlockMasterClient;
+import tachyon.client.UserMasterClient;
+import tachyon.client.ClientContext;
+import tachyon.thrift.NetAddress;
+import tachyon.thrift.WorkerInfo;
+import tachyon.util.ThreadFactoryUtils;
+import tachyon.util.network.NetworkAddressUtils;
+import tachyon.worker.ClientMetrics;
+import tachyon.worker.WorkerClient;
+
+/**
+ * A shared context in each client JVM for common Block Store client functionality such as a pool
+ * of master clients and a pool of local worker clients. Any remote clients will be created and
+ * destroyed on a per use basis.
+ */
+public enum BSContext {
+  INSTANCE;
+
+  private BlockMasterClientPool mBlockMasterClientPool;
+  private BlockWorkerClientPool mLocalBlockWorkerClientPool;
+  private final ExecutorService mRemoteBlockWorkerExecutor;
+
+  private BSContext() {
+    mBlockMasterClientPool =
+        new BlockMasterClientPool(ClientContext.getMasterAddress(), ClientContext.getConf());
+    // TODO: Get size from configuration
+    mRemoteBlockWorkerExecutor =
+        Executors.newFixedThreadPool(10,
+            ThreadFactoryUtils.build("remote-block-worker-heartbeat-%d", true));
+
+    NetAddress localWorkerAddress =
+        getWorkerAddress(NetworkAddressUtils.getLocalHostName(ClientContext.getConf()));
+
+    // If the local worker is not available, do not initialize the local worker client pool
+    if (null == localWorkerAddress) {
+      mLocalBlockWorkerClientPool = null;
+    } else {
+      mLocalBlockWorkerClientPool =
+          new BlockWorkerClientPool(localWorkerAddress, ClientContext.getConf());
+    }
+  }
+
+  /**
+   * Reinitializes the Block Store context. This method should only be used in ClientContext.
```

Comment:
There is a programmatic way to enforce that this method is only use in ClientContext: http://stackoverflow.com/questions/182278/is-there-a-way-to-simulate-the-c-friend-concept-in-java